### ISSUE TITLE
Bug: simultaneously uploading multiple datasets with tm_client fails

### ISSUE THREAD

**Issue by scottberry** [pelkmanslab/TissueMAPS-OLD#21]
When trying to upload data via tm_client to two different experiments on two different virtual machines (say VM-A and VM-B), the upload fails because tm_client seems to confuse the list of expected files between these uploads. The error message is something like "fileXXX.png cannot be found, path does not exist". This is because fileXXX.png does not exist on that virtual machine (e.g. tm_client running on VM-A is looking for a filename in the list of expected files on VM-B). I guess the problem is that expected uploads is somehow linked to "username" but not specific between different experiments.

At the moment, it is therefore not possible to simultaneously upload two datasets, but eventually we would like to do this (even for different acquisitions or plates in the same experiment)

**Comment by hackermd** [comment]
Could you please provide the exact commands (excluding your password) and the respective log outputs?

**Comment by hackermd** [comment]
`tm_client` creates a separate `tmclient.api.TmClient` instance for each call and the `directory` provided to `tmclient.api.TmClient.upload_microscope_files()` is handled independently and excepted to be on the machine that calls `tm_client` program.

**Comment by scottberry** [comment]
This was a while ago, but I only just got around to adding it to the issues. I’ll try to generate an example case.

**Comment by scottberry** [comment]
I haven't yet been able to recreate the issue but should I be concerned that two simultaneous uploads to different experiments have the same acquisition (dG1hcHMx)?

Is this because the acquisitions have the same name?

Executed on virtual machine #1:

tm_client -vv -H chimp.tissuemaps.org -u scottberry -p $PASS -e 20170106_CytooStarter1 upload -d ~/data/Data/Users/Scott/Cytoo/20170106_CytooStarter1/ microscope_file -p Starter1 -a DAPI-SEA647 > uploadCytoo1.log 2>&1 &

2017-02-02 08:03:38 | INFO     | tmclient.api                             | upload microscope files of plate "Starter1" and acquisition "DAPI-SEA647"
2017-02-02 08:03:38 | INFO     | tmclient.api                             | register files for upload of acquisition dG1hcHMx

Executed on virtual machine #2:

tm_client -vv -H chimp.tissuemaps.org -u scottberry -p $PASS -e Cytoo_Starter2_MIP upload -d ~/data/Data/Users/Scott/Cytoo/MIP_test/ microscope_file -p MIP -a DAPI-SEA647 > upload.log 2>&1 &

2017-02-02 11:55:40 | INFO     | tmclient.api                             | upload microscope files of plate "MIP" and acquisition "DAPI-SEA647"
2017-02-02 11:55:41 | INFO     | tmclient.api                             | register files for upload of acquisition dG1hcHMx


Scott

**Comment by hackermd** [comment]
> I haven't yet been able to recreate the issue but should I be concerned that two simultaneous uploads to different experiments have the same acquisition (dG1hcHMx)?

No! `dG1hcHMx` is the (hashed) acquisition ID. This ID is experiment-specific. The combination of experiment ID and acquisition ID must be unique. This is ensured by two mechanisms:
- each experiment resides in a separate database
- experiment names must be unique per user (ensured by a `UNIQUE CONSTRAINT` on table "experiment_references" in the main database)

So the combination of user, experiment and acquisition name is unique.

> Is this because the acquisitions have the same name?

This is the (hashed) acquisition ID assigned within the experiment-specific database. So if two experiments have only one acquisition, both acquisitions will have the same ID.

When you look at the route for the upload, you'll notice that a combination of `experiment_id` and `acquisition_id` is required:
`/experiments/<experiment_id>/acquisitions/<acquisition_id>/upload/upload-file
`

You can also see the actual HTTP request when increasing logging verbosity: `tm_client -vvv`

**Comment by scottberry** [comment]
Ok. If you would like to close the issue, please go ahead. I’ll refer to it again later if I have the same problem again.
Thanks,
Scott

**Comment by scottberry** [comment]
I experienced this issue again on an instance from image tissuemaps-v0.2.0[dev-20170214] 

Here is the screen dump of the error:

```
2017-02-15 22:27:44 | DEBUG    | tmclient.base                            | url: http://172.23.92.99:80/api/experiments/dG1hcHMy/acquisitions/dG1hcHMx/upload/upload-file
2017-02-15 22:27:44 | INFO     | tmclient.api                             | upload file: AssayPlate_Greiner_655090_D03_T0001F032L01A01Z29C02.tif
2017-02-15 22:27:44 | DEBUG    | tmclient.api                             | upload file "/home/ubuntu/data/Data/Users/Scott/20170111-184A1-NascentRNACycler-Stack/AssayPlate_Greiner_655090_D03_T0001F032L01A01Z29C02.tif" for acquisition dG1hcHMx
2017-02-15 22:27:44 | DEBUG    | tmclient.base                            | url: http://172.23.92.99:80/api/experiments/dG1hcHMy/acquisitions/dG1hcHMx/upload/upload-file
2017-02-15 22:27:45 | INFO     | tmclient.api                             | upload file: AssayPlate_Greiner_655090_D03_T0001F060L01A01Z08C02.tif
2017-02-15 22:27:45 | DEBUG    | tmclient.api                             | upload file "/home/ubuntu/data/Data/Users/Scott/20170111-184A1-NascentRNACycler-Stack/AssayPlate_Greiner_655090_D03_T0001F060L01A01Z08C02.tif" for acquisition dG1hcHMx
2017-02-15 22:27:45 | DEBUG    | tmclient.base                            | url: http://172.23.92.99:80/api/experiments/dG1hcHMy/acquisitions/dG1hcHMx/upload/upload-file
Traceback (most recent call last):
  File "/usr/local/bin/tm_client", line 9, in <module>
    load_entry_point('tmclient==0.0.1', 'console_scripts', 'tm_client')()
  File "build/bdist.linux-x86_64/egg/tmclient/api.py", line 108, in __main__
  File "build/bdist.linux-x86_64/egg/tmclient/api.py", line 91, in __call__
  File "build/bdist.linux-x86_64/egg/tmclient/api.py", line 725, in upload_microscope_files
  File "build/bdist.linux-x86_64/egg/tmclient/api.py", line 775, in _upload_file
  File "/usr/local/lib/python2.7/dist-packages/requests-2.12.4-py2.7.egg/requests/models.py", line 893, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: BAD REQUEST for url: http://172.23.92.99:80/api/experiments/dG1hcHMy/acquisitions/dG1hcHMx/upload/upload-file
```
and after a bit of digging @HackerMD found also the following info

```
2017-02-15 22:38:06 | ERROR    | wsgi                                     | Exception on /api/experiments/dG1hcHMy/acquisitions/dG1hcHMx/upload/upload-file [POST]
Traceback (most recent call last):
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask_jwt/__init__.py", line 177, in decorator
    return fn(*args, **kwargs)
  File "/home/ubuntu/tmserver/tmserver/util.py", line 207, in wrapped
    return f(*args, **kwargs)
  File "/home/ubuntu/tmserver/tmserver/api/upload.py", line 233, in upload_file
    acquisition = session.query(tm.Acquisition).get(acquisition_id)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 831, in get
    return self._get_impl(ident, loading.load_on_ident)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 864, in _get_impl
    return fallback_fn(self, key)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/loading.py", line 223, in load_on_ident
    return q.one()
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2749, in one
    ret = self.one_or_none()
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2719, in one_or_none
    ret = list(self)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2790, in __iter__
    return self._execute_and_instances(context)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2813, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 945, in execute
    return meth(self, multiparams, params)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 263, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1053, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1189, in _execute_context
    context)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1393, in _handle_dbapi_exception
    exc_info
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
    context)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 470, in do_execute
    cursor.execute(statement, parameters)
ProgrammingError: (psycopg2.ProgrammingError) relation "experiment_3.acquisitions" does not exist
LINE 2: FROM experiment_3.acquisitions
             ^
 [SQL: 'SELECT experiment_3.acquisitions.location AS acquisitions_location, experiment_3.acquisitions.id AS acquisitions_id, experiment_3.acquisitions.name AS acquisitions_name, experiment_3.acquisitions.description AS acquisitions_description, experiment_3.acquisitions.plate_id AS acquisitions_plate_id, experiment_3.acquisitions.created_at AS acquisitions_created_at, experiment_3.acquisitions.updated_at AS acquisitions_updated_at \nFROM experiment_3.acquisitions \nWHERE experiment_3.acquisitions.id = %(param_1)s'] [parameters: {'param_1': 1}]
[pid: 1156|app: 0|req: 73/1632] 172.23.82.3 () {42 vars in 948 bytes} [Wed Feb 15 22:38:06 2017] POST /api/experiments/dG1hcHMy/acquisitions/dG1hcHMx/upload/upload-file => generated 291 bytes in 45 msecs (HTTP/1.1 500) 2 headers in 84 bytes (6 switches on core 99)
```

This may also be related to issue #23

**Comment by scottberry** [comment]
I don't know the details of how this works but I think there is something wrong with the upload “register”. I just had another case of the same error and after restarting the server this error seems to DISAPPEAR.

That is, whatever causes the uploader to look for files which were previously on the upload list is cleared upon server restart.

**Comment by hackermd** [comment]
I don't understand. Could you elaborate?

The client keeps a list of files that should be uploaded (those that were dropped) and tries to upload each file individually. The server first registers all files that should be uploaded in the database and keeps track of their status. When he handles an upload request, he checks whether the file has already been uploaded, i.e. queries the state of the file in the database. Restarting the server shouldn't have an impact on either the client side list nor on the files registered in the database.

**Comment by scottberry** [comment]
I am not sure what the problem is, but you see from the logs that after happily uploading a couple of images to experiment_5, the server tries to get an image from experiment_3, which no longer exists (as I deleted it). Perhaps on server restart the experiment is detected as "deleted"?

```
2017-02-16 14:07:51 | INFO     | tmserver.api.experiment                  | create experiment "20170111-184A1-NascentRNACycler-Stack2"
[pid: 4715|app: 0|req: 1/35] 130.60.76.122 () {48 vars in 1024 bytes} [Thu Feb 16 14:07:51 2017] POST /api/experiments => generated 136 bytes in 22463 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
2017-02-16 14:08:13 | INFO     | tmserver.api.experiment                  | get experiments
[pid: 4715|app: 0|req: 2/36] 130.60.76.122 () {42 vars in 885 bytes} [Thu Feb 16 14:08:13 2017] GET /api/experiments => generated 437 bytes in 7 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
2017-02-16 14:08:20 | INFO     | tmserver.api.experiment                  | get experiments
[pid: 4715|app: 0|req: 3/37] 130.60.76.122 () {42 vars in 885 bytes} [Thu Feb 16 14:08:20 2017] GET /api/experiments => generated 437 bytes in 6 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
2017-02-16 14:08:22 | INFO     | tmserver.api.experiment                  | delete experiment 3
2017-02-16 14:08:22 | INFO     | tmlib.models.utils                       | drop schema of experiment 3
[pid: 4716|app: 0|req: 2/38] 130.60.76.122 () {44 vars in 940 bytes} [Thu Feb 16 14:08:22 2017] DELETE /api/experiments/dG1hcHMz => generated 22 bytes in 3094 msecs (HTTP/1.1 200) 2 headers in 71 bytes (4 switches on core 99)
[pid: 4714|app: 0|req: 5/39] 172.23.82.3 () {40 vars in 509 bytes} [Thu Feb 16 14:08:47 2017] POST /auth => generated 212 bytes in 326 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
2017-02-16 14:08:47 | INFO     | tmserver.api.experiment                  | get experiments
2017-02-16 14:08:47 | INFO     | tmserver.api.experiment                  | filter experiments for name "20170111-184A1-NascentRNACycler-Stack2"
[pid: 4711|app: 0|req: 6/40] 172.23.82.3 () {38 vars in 750 bytes} [Thu Feb 16 14:08:47 2017] GET /api/experiments?name=20170111-184A1-NascentRNACycler-Stack2 => generated 165 bytes in 7 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
2017-02-16 14:08:47 | INFO     | tmserver.api.experiment                  | create plate "test3D" for experiment 5
[pid: 4707|app: 0|req: 7/41] 172.23.82.3 () {42 vars in 776 bytes} [Thu Feb 16 14:08:47 2017] POST /api/experiments/dG1hcHM1/plates => generated 162 bytes in 20 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
[pid: 4707|app: 0|req: 8/42] 172.23.82.3 () {40 vars in 509 bytes} [Thu Feb 16 14:08:48 2017] POST /auth => generated 212 bytes in 318 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
2017-02-16 14:08:48 | INFO     | tmserver.api.experiment                  | get experiments
2017-02-16 14:08:48 | INFO     | tmserver.api.experiment                  | filter experiments for name "20170111-184A1-NascentRNACycler-Stack2"
[pid: 4707|app: 0|req: 9/43] 172.23.82.3 () {38 vars in 750 bytes} [Thu Feb 16 14:08:48 2017] GET /api/experiments?name=20170111-184A1-NascentRNACycler-Stack2 => generated 165 bytes in 5 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
2017-02-16 14:08:48 | INFO     | tmserver.api.experiment                  | create acquisition "test3D" for plate "test3D" from experiment 5
[pid: 4709|app: 0|req: 1/44] 172.23.82.3 () {42 vars in 788 bytes} [Thu Feb 16 14:08:48 2017] POST /api/experiments/dG1hcHM1/acquisitions => generated 114 bytes in 122 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
[pid: 4709|app: 0|req: 2/45] 172.23.82.3 () {40 vars in 509 bytes} [Thu Feb 16 14:08:48 2017] POST /auth => generated 212 bytes in 302 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
2017-02-16 14:08:49 | INFO     | tmserver.api.experiment                  | get experiments
2017-02-16 14:08:49 | INFO     | tmserver.api.experiment                  | filter experiments for name "20170111-184A1-NascentRNACycler-Stack2"
[pid: 4719|app: 0|req: 2/46] 172.23.82.3 () {38 vars in 750 bytes} [Thu Feb 16 14:08:49 2017] GET /api/experiments?name=20170111-184A1-NascentRNACycler-Stack2 => generated 165 bytes in 7 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
2017-02-16 14:08:49 | INFO     | tmserver.api.experiment                  | get acquistions for experiment 5
[pid: 4710|app: 0|req: 5/47] 172.23.82.3 () {38 vars in 766 bytes} [Thu Feb 16 14:08:49 2017] GET /api/experiments/dG1hcHM1/acquisitions?plate_name=test3D&name=test3D => generated 134 bytes in 11 msecs (HTTP/1.1 200) 2 headers in 72 bytes (4 switches on core 99)
[pid: 4711|app: 0|req: 7/48] 172.23.82.3 () {42 vars in 846 bytes} [Thu Feb 16 14:08:49 2017] POST /api/experiments/dG1hcHM1/acquisitions/dG1hcHMx/upload/register => generated 839756 bytes in 3481 msecs (HTTP/1.1 200) 2 headers in 75 bytes (9 switches on core 99)
2017-02-16 14:08:52 | INFO     | tmserver.api.upload                      | upload file "AssayPlate_Greiner_655090_D03_T0001F027L01A01Z33C01.tif"
[pid: 4711|app: 0|req: 8/49] 172.23.82.3 () {42 vars in 948 bytes} [Thu Feb 16 14:08:52 2017] POST /api/experiments/dG1hcHM1/acquisitions/dG1hcHMx/upload/upload-file => generated 22 bytes in 73 msecs (HTTP/1.1 200) 2 headers in 71 bytes (59 switches on core 99)
2017-02-16 14:08:53 | INFO     | tmserver.api.upload                      | upload file "AssayPlate_Greiner_655090_E03_T0001F042L01A01Z23C02.tif"
[pid: 4711|app: 0|req: 9/50] 172.23.82.3 () {42 vars in 948 bytes} [Thu Feb 16 14:08:52 2017] POST /api/experiments/dG1hcHM1/acquisitions/dG1hcHMx/upload/upload-file => generated 22 bytes in 89 msecs (HTTP/1.1 200) 2 headers in 71 bytes (9 switches on core 99)
2017-02-16 14:08:53 | ERROR    | wsgi                                     | Exception on /api/experiments/dG1hcHM1/acquisitions/dG1hcHMx/upload/upload-file [POST]
Traceback (most recent call last):
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask/app.py", line 1982, in wsgi_app
    response = self.full_dispatch_request()
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask/app.py", line 1614, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask/app.py", line 1517, in handle_user_exception
    reraise(exc_type, exc_value, tb)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask/app.py", line 1612, in full_dispatch_request
    rv = self.dispatch_request()
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask/app.py", line 1598, in dispatch_request
    return self.view_functions[rule.endpoint](**req.view_args)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/flask_jwt/__init__.py", line 177, in decorator
    return fn(*args, **kwargs)
  File "/home/ubuntu/tmserver/tmserver/util.py", line 207, in wrapped
    return f(*args, **kwargs)
  File "/home/ubuntu/tmserver/tmserver/api/upload.py", line 233, in upload_file
    acquisition = session.query(tm.Acquisition).get(acquisition_id)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 831, in get
    return self._get_impl(ident, loading.load_on_ident)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 864, in _get_impl
    return fallback_fn(self, key)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/loading.py", line 223, in load_on_ident
    return q.one()
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2749, in one
    ret = self.one_or_none()
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2719, in one_or_none
    ret = list(self)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2790, in __iter__
    return self._execute_and_instances(context)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/orm/query.py", line 2813, in _execute_and_instances
    result = conn.execute(querycontext.statement, self._params)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 945, in execute
    return meth(self, multiparams, params)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/sql/elements.py", line 263, in _execute_on_connection
    return connection._execute_clauseelement(self, multiparams, params)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1053, in _execute_clauseelement
    compiled_sql, distilled_params
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1189, in _execute_context
    context)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1393, in _handle_dbapi_exception
    exc_info
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/util/compat.py", line 203, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/base.py", line 1182, in _execute_context
    context)
  File "/home/ubuntu/.local/lib/python2.7/site-packages/sqlalchemy/engine/default.py", line 470, in do_execute
    cursor.execute(statement, parameters)
ProgrammingError: (psycopg2.ProgrammingError) relation "experiment_3.acquisitions" does not exist
LINE 2: FROM experiment_3.acquisitions
             ^
 [SQL: 'SELECT experiment_3.acquisitions.location AS experiment_3_acquisitions_location, experiment_3.acquisitions.id AS experiment_3_acquisitions_id, experiment_3.acquisitions.name AS acquisitions_name, experiment_3.acquisitions.description AS experiment_3_acquisitions_description, experiment_3.acquisitions.plate_id AS acquisitions_plate_id, experiment_3.acquisitions.created_at AS experiment_3_acquisitions_created_at, experiment_3.acquisitions.updated_at AS experiment_3_acquisitions_updated_at \nFROM experiment_3.acquisitions \nWHERE experiment_3.acquisitions.id = %(param_1)s'] [parameters: {'param_1': 1}]
[pid: 4706|app: 0|req: 4/51] 172.23.82.3 () {42 vars in 948 bytes} [Thu Feb 16 14:08:53 2017] POST /api/experiments/dG1hcHM1/acquisitions/dG1hcHMx/upload/upload-file => generated 291 bytes in 131 msecs (HTTP/1.1 500) 2 headers in 84 bytes (59 switches on core 99)
Brutally killing worker 14 (pid: 4718)...
Brutally killing worker 8 (pid: 4712)...
Thu Feb 16 14:13:05 2017 - stopping gevent signals watchers for worker 14 (pid: 4718)...
Thu Feb 16 14:13:05 2017 - stopping gevent signals watchers for worker 8 (pid: 4712)...
Thu Feb 16 14:13:05 2017 - stopping gevent sockets watchers for worker 14 (pid: 4718)...
Thu Feb 16 14:13:05 2017 - stopping gevent sockets watchers for worker 8 (pid: 4712)...
```

**Comment by scottberry** [comment]
By the way ```Brutally killing worker 14``` is a quite tough...

**Comment by hackermd** [comment]
I don't know what's going on. I assume it has to do with the `session` and retrieval of the correct `experiment_id` in the `tmserver` app upon decoding of URLs.

I recently upgraded to newer versions of `sqlalchemy` and `psycopg2`.

**Comment by hackermd** [comment]
This was a tricky one, but I think it's fixed with TissueMAPS/TmLibrary@ce86a5659c88e092f5478f4316b57385120436e4.

This is my theory of what happened:
We use a separate [schema](https://www.postgresql.org/docs/current/static/ddl-schemas.html) for each experiment using the following naming convention: `experiment_<id>.<table_name>` - e.g. `experiment_2.acquisitions` as in the above error traceback. Tables can be queried in two ways:
1) Using the full name: 

```sql
SELECT * FROM experiment_2.acquisitions;
```

2) Setting `search_path` and using only the relative name:

```sql
SET search_path TO 'experiment_2';
SELECT * FROM acquisitions;
```

I implemented the second approach in `tmlib.models.utils.ExperimentSession` , such that `search_path` get's automatically set when the database connection is checked out from the connection pool. I also added event listeners that would set `search_path` again in case the connection was returned to the pool and a new one was checked out. I added these listeners at the wrong location, however. So in rare cases when a new connection was checked out from the pool with a wrong `search_path` the table could not be found.

**Comment by hackermd** [comment]
This didn't totally fixed it! The `search_path` did get set at wrong time. SQLAlchmemy sometimes makes things harder than they have to be!!

It is now hopefully fixed with TissueMAPS/TmLibrary@d3fe2ea3ba91824f228f8474b61d1452600dbb6d.
The `search_path` gets now only set after `BEGIN` or `ROLLBACK` and not after `COMMIT`.

**Comment by hackermd** [comment]
After further digging I came to the conclusion that the error is most likely not due to an incorrect `search_path` setting, but rather due to the fact that the schema gets set on the table for table creation. This can cause problems later on, because the old schema settings can get mingled up with new queries that set the schema via `search_path`.

It seems that this can be solved by removing the schema settings from the table after tables have been created, see [here](https://github.com/TissueMAPS/TmLibrary/blob/72343cf0a4d47d2fd22d6808cee78cb3ebcb1e20/tmlib/models/utils.py#L182). However, the problem could still occur in case several users create tables at the same time. Using a copy of the table seems to solve this issue, see [here](https://github.com/TissueMAPS/TmLibrary/blob/3904e2bf651e4fac53fbe6d3094c968433448364/tmlib/models/utils.py#L193).
